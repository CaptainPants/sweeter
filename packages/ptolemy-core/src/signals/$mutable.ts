import { MutableValueSignal } from './internal/Signal-implementations/MutableValueSignal.js';
import { type ReadWriteSignal } from './types.js';

// The normal version must be before the optional version for priority reasons.

/**
 * Create a new ReadWriteSignal signal with an initial value.
 * @param initialValue
 * @returns
 */
export function $mutable<T>(initialValue: T): ReadWriteSignal<T>;
/**
 * Create a new ReadWriteSignal signal with a potentially undefined initial value.
 * @param initialValue
 * @returns
 */
export function $mutable<T>(initialValue?: T): ReadWriteSignal<T | undefined>;
/**
 * Create a new ReadWriteSignal signal with an initial value.
 * @param initialValue
 * @returns
 */
export function $mutable<T>(initialValue?: T): ReadWriteSignal<T> {
    if (arguments.length === 0) {
        // arguments.length == 0 means the optional parameter was not passed
        return new MutableValueSignal<T>();
    }
    return new MutableValueSignal<T>(initialValue);
}
