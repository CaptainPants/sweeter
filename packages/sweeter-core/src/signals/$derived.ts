import { MutableCalculatedSignal } from './internal/MutableCalculatedSignal.js';
import { type ReadWriteSignal } from './types.js';

export function $derived<TSource, TKey extends keyof TSource>(
    source: ReadWriteSignal<TSource>,
    key: TKey,
): ReadWriteSignal<TSource[TKey]> & { value: TSource[TKey] } {
    return new MutableCalculatedSignal<TSource[TKey]>(
        () => source.value[key],
        (value) => {
            const sourceObjectOriginal = source.value;

            if (Array.isArray(sourceObjectOriginal)) {
                if (typeof key !== 'number')
                    throw new TypeError(
                        'If source is an array, key must be a number.',
                    );

                const copy = [...sourceObjectOriginal] as TSource;
                copy[key as TKey] = value;
                source.update(copy);
            } else if (
                typeof sourceObjectOriginal === 'object' &&
                sourceObjectOriginal !== null
            ) {
                const copy = { ...sourceObjectOriginal } as TSource;
                copy[key] = value;
                source.update(copy);
            } else {
                throw new TypeError(
                    `Unexpected type ${typeof sourceObjectOriginal}`,
                );
            }
        },
    );
}
