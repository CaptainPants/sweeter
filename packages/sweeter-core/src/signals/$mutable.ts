import { MutableValueSignal } from './internal/MutableValueSignal.js';
import { type ReadWriteSignal } from './types.js';

/**
 * Create a new ReadWriteSignal signal with an initial value.
 * @param initialValue
 * @returns
 */
export function $mutable<T>(initialValue: T): ReadWriteSignal<T> & { value: T };
/**
 * Create a new ReadWriteSignal signal with an undefined initial value.
 * @param initialValue
 * @returns
 */
export function $mutable<TOrUndefined>(): ReadWriteSignal<
    TOrUndefined | undefined
> & { value: TOrUndefined extends undefined ? never : TOrUndefined };
export function $mutable<T>(
    initialValue?: T,
): ReadWriteSignal<T> & { value: T } {
    return new MutableValueSignal<T>(initialValue as T);
}
