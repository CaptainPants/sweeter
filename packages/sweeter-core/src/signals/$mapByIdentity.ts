import { type Signal } from '../signals/types.js';
import { type MightBeSignal } from '../types.js';
import { isSignal } from './isSignal.js';
import { $derived } from './$derived.js';
import { $peek, $subscribe, $val } from './$val.js';
import { $constant } from './$constant.js';
import { createIdentityCache } from './internal/createIdentityCache.js';

/**
 * Create a new signal by mapping each element of an input signal, using a mapping function.
 * @param items
 * @param mappingFun
 * @param orderBy
 * @returns
 */
export function $mapByIdentity<TInput, TMapped>(
    items: MightBeSignal<readonly TInput[]>,
    mappingFun: MightBeSignal<(item: TInput, index: Signal<number>) => TMapped>,
    orderBy: MightBeSignal<(obj: TMapped, source: TInput) => string | number>,
): Signal<readonly TMapped[]> {
    const cache = createIdentityCache<TInput, TMapped>();

    if (isSignal(mappingFun)) {
        cache.clearOnChange(mappingFun); // If the mapping function changes, clear the cache
    }
    if (isSignal(orderBy)) {
        cache.clearOnChange(orderBy); // If the orderBy function changes, clear the cache
    }

    // If the input signal changes, remove any items that are no longer in the input
    // This may also be triggered by the mappingFun or orderBy value changing, which will also
    // trigger a clear to happen, prior to the fixUpIndexes
    cache.updateOnChange(items, mappingFun, orderBy);
    cache.update($peek(items), $peek(mappingFun), $peek(orderBy));

    const resultSignal = $derived(() => {
        $subscribe(mappingFun); // If the mapping function changes, the whole cache is invalidated, and we need to load our data again
        $subscribe(orderBy);
        $subscribe(items);

        const cacheItems = cache.getItems();

        return cacheItems.map((x) => x.mappedElement);
    });

    cache.keepAliveWith(resultSignal); // Keep the cache object alive as long as resultSignal is alive

    return resultSignal;
}
