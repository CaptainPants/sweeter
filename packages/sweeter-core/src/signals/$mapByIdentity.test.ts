import { $derived } from "./$derived";
import { $mapByIdentity } from "./$mapByIdentity";
import { $mutable } from "./$mutable";

it('Does stuff', () => {
    const input = $mutable<string[]>(['alpha', 'beta']);

    const mapped = $mapByIdentity(
        input,
        (input, index) => {
            return $derived(() => input + '_TEST_' + index.value);
        },
        (item, source) => source
    );

    expect(mapped.peek().map(x => x.peek())).toMatchSnapshot();

    input.value = ['beta', 'gamma'];

    expect(mapped.peek().map(x => x.peek())).toMatchSnapshot();
});