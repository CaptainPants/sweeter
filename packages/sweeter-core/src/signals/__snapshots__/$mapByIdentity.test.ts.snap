// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Does stuff 1`] = `
[
  "alpha_TEST_0",
  "beta_TEST_1",
]
`;

exports[`Does stuff 2`] = `
[
  "beta_TEST_0",
  "gamma_TEST_1",
]
`;
